File: java/355.design-twitter.java
/*
 * @lc app=leetcode id=355 lang=java
 *
 * [355] Design Twitter
 */

// @lc code=start
class Tweet {
    public int index;
    public int userId;
    public int timestamp;
    public int tweetId;

    public Tweet(int index, int userId, int timestamp, int tweetId) { 
        this.index = index;
        this.userId = userId;
        this.timestamp = timestamp;
        this.tweetId = tweetId;
    }
}

class Twitter {

    private Map<Integer, List<Tweet>> tweets;
    private int clock;
    private Map<Integer, Set<Integer>> follows;

    public Twitter() {
        tweets = new HashMap();
        clock = 0;
        follows = new HashMap();
    }
    
    public void postTweet(int userId, int tweetId) {
        List<Tweet> userTweets = this.tweets.get(userId);
        if(userTweets == null) {
            userTweets = new ArrayList();
            this.tweets.put(userId, userTweets);
        }

        userTweets.add(new Tweet(userTweets.size(), userId, this.clock, tweetId));
        this.clock++;
    }
    
    public List<Integer> getNewsFeed(int userId) {
        Set<Integer> userFollows = this.follows.get(userId);

        PriorityQueue<Tweet> pq = new PriorityQueue<Tweet>(Comparator.comparingInt(t -> -1 * t.timestamp));

        List<Tweet> userTweets = this.tweets.get(userId);
        if(userTweets != null) {
            pq.add(userTweets.get(userTweets.size() - 1));
        }

        if(userFollows != null) {
            for(Integer followee: userFollows) {
                userTweets = this.tweets.get(followee);
                if(userTweets == null) continue;

                pq.add(userTweets.get(userTweets.size() - 1));
            }
        }

        List<Integer> res = new ArrayList();
        while(res.size() < 10 && !pq.isEmpty()) {
            Tweet t = pq.poll();

            res.add(t.tweetId);
            
            if(t.index == 0) continue;

            pq.add(this.tweets.get(t.userId).get(t.index - 1));
        }

        return res;
    }
    
    public void follow(int followerId, int followeeId) {
        Set<Integer> userFollows = this.follows.get(followerId);
        if(userFollows == null) {
            userFollows = new HashSet();
            this.follows.put(followerId, userFollows);
        }

        userFollows.add(followeeId);
    }
    
    public void unfollow(int followerId, int followeeId) {
        Set<Integer> userFollows = this.follows.get(followerId);
        if(userFollows == null) {
            return;
        }

        userFollows.remove(followeeId);
    }
}
/**
 * Your Twitter object will be instantiated and called as such:
 * Twitter obj = new Twitter();
 * obj.postTweet(userId,tweetId);
 * List<Integer> param_2 = obj.getNewsFeed(userId);
 * obj.follow(followerId,followeeId);
 * obj.unfollow(followerId,followeeId);
 */
// @lc code=end

File: cpp/355.design-twitter.cpp
/*
 * @lc app=leetcode id=355 lang=cpp
 *
 * [355] Design Twitter
 */

// @lc code=start
struct Tweet {
    int timestamp;
    int tweetId;
    int userId;
};

class Twitter {
public:
    Twitter() {
        
    }
    
    void postTweet(int userId, int tweetId) {
        posts_map_[userId].push_back({timestamp_, tweetId, userId});
        timestamp_++;
    }
    
    vector<int> getNewsFeed(int userId) {
        using T = vector<Tweet>::reverse_iterator;
        auto compare = [](T a, T b) {return a->timestamp < b->timestamp;};
        priority_queue<T, vector<T>, decltype(compare)> pq(compare);

        if(posts_map_[userId].rbegin() != posts_map_[userId].rend()) {
            pq.push(posts_map_[userId].rbegin());
        }

        for(int followeeId : following_map_[userId]) {
            if(posts_map_[followeeId].rbegin() != posts_map_[followeeId].rend()) {
                pq.push(posts_map_[followeeId].rbegin());
            }
        }

        vector<int> res;

        int k = 10;
        while(k-- && !pq.empty()) {
            auto it = pq.top();
            res.push_back(it->tweetId);
            pq.pop();
            if(it + 1 != posts_map_[it->userId].rend()) {
                pq.push(it + 1);
            }
        }

        return res;
    }
    
    void follow(int followerId, int followeeId) {
        following_map_[followerId].insert(followeeId);
    }
    
    void unfollow(int followerId, int followeeId) {
        following_map_[followerId].erase(followeeId);
    }
private:
    int timestamp_ = 0;
    unordered_map<int, unordered_set<int>> following_map_;
    unordered_map<int, vector<Tweet>> posts_map_;
};

/**
 * Your Twitter object will be instantiated and called as such:
 * Twitter* obj = new Twitter();
 * obj->postTweet(userId,tweetId);
 * vector<int> param_2 = obj->getNewsFeed(userId);
 * obj->follow(followerId,followeeId);
 * obj->unfollow(followerId,followeeId);
 */
// @lc code=end

